src/parcsr_ls/par_relax_more.c

```diff
@@ -328,275 +418,26 @@ means half, and .1 means 10percent)
-   }
-   else /* scaling! */
-   {
-      /*grab 1/sqrt(diagonal) */
-
-      ds = hypre_ParVectorCreate(hypre_ParCSRMatrixComm(A),
-                                 hypre_ParCSRMatrixGlobalNumRows(A),
-                                 hypre_ParCSRMatrixRowStarts(A));
-      hypre_ParVectorInitialize(ds);
-      hypre_ParVectorSetPartitioningOwner(ds,0);
-      ds_data = hypre_VectorData(hypre_ParVectorLocalVector(ds));
-
-      tmp_vec = hypre_ParVectorCreate(hypre_ParCSRMatrixComm(A),
-                                      hypre_ParCSRMatrixGlobalNumRows(A),
-                                      hypre_ParCSRMatrixRowStarts(A));
-      hypre_ParVectorInitialize(tmp_vec);
-      hypre_ParVectorSetPartitioningOwner(tmp_vec,0);
-      tmp_data = hypre_VectorData(hypre_ParVectorLocalVector(tmp_vec));
-
-      /* get ds_data and get scaled residual: r = D^(-1/2)f -
-       * D^(-1/2)A*u */
-#ifdef HYPRE_USING_OPENMP
-#pragma omp parallel for private(j,diag) HYPRE_SMP_SCHEDULE
-#endif
-      for (j = 0; j < num_rows; j++)
-      {
-         diag = A_diag_data[A_diag_i[j]];
-         ds_data[j] = 1/sqrt(diag);
-
-         r_data[j] = ds_data[j] * f_data[j];
-      }
-
-      hypre_ParCSRMatrixMatvec(-1.0, A, u, 0.0, tmp_vec);
-
-#ifdef HYPRE_USING_OPENMP
-#pragma omp parallel for private(j) HYPRE_SMP_SCHEDULE
-#endif
-      for ( j = 0; j < num_rows; j++ )
-      {
-         r_data[j] += ds_data[j] * tmp_data[j];
-      }
-
-      /* save original u, then start
-         the iteration by multiplying r by the cheby coef.*/
-#ifdef HYPRE_USING_OPENMP
-#pragma omp parallel for private(j) HYPRE_SMP_SCHEDULE
-#endif
-      for ( j = 0; j < num_rows; j++ )
-      {
-         orig_u[j] = u_data[j]; /* orig, unscaled u */
-
-         u_data[j] = r_data[j] * coefs[cheby_order];
-      }
-
-      /* now do the other coefficients */
-      for (i = cheby_order - 1; i >= 0; i-- )
-      {
-         /* v = D^(-1/2)AD^(-1/2)u */
-#ifdef HYPRE_USING_OPENMP
-#pragma omp parallel for private(j) HYPRE_SMP_SCHEDULE
-#endif
-         for ( j = 0; j < num_rows; j++ )
-         {
-            tmp_data[j]  =  ds_data[j] * u_data[j];
-         }
-         hypre_ParCSRMatrixMatvec(1.0, A, tmp_vec, 0.0, v);
-
-         /* u_new = coef*r + v*/
-         mult = coefs[i];
-
-#ifdef HYPRE_USING_OPENMP
-#pragma omp parallel for private(j,tmp_d) HYPRE_SMP_SCHEDULE
-#endif
-         for ( j = 0; j < num_rows; j++ )
-         {
-            tmp_d = ds_data[j]* v_data[j];
-            u_data[j] = mult * r_data[j] + tmp_d;
-         }
-
-      } /* end of cheby_order loop */
-
-      /* now we have to scale u_data before adding it to u_orig*/
-
-#ifdef HYPRE_USING_OPENMP
-#pragma omp parallel for private(j) HYPRE_SMP_SCHEDULE
-#endif
-      for ( j = 0; j < num_rows; j++ )
-      {
-         u_data[j] = orig_u[j] + ds_data[j]*u_data[j];
-      }
-
-      hypre_ParVectorDestroy(ds);
-      hypre_ParVectorDestroy(tmp_vec);
-
+   HYPRE_Real *coefs   = NULL;
+   HYPRE_Real *ds_data = NULL;
```

Comment:
It does Setup in each time calling this function from `ams.c`. We should change it.